### PR TITLE
Add missing inner IP header in ICMP error-reply packet

### LINKED ISSUE #21236: dsr_reply_icmp4 does not add original ip header in the icmp message
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I have enabled DSR as well as ENABLE_DSR_ICMP_ERRORS, so that icmp error packet is returned to the sender in case DSR packet becomes bigger than the MTU size.

UDP client sends a UDP packet of size equal to MTU of the lb node. The packet is destined to UDP service. When the packet reaches lb node, DSR code kicks in and adds 8 bytes of additional data. This makes the packet size larger than MTU and lb node usually just drops it. Since ENABLE_DSR_ICMP_ERRORS is enabled, dsr_reply_icmp4() sends the response back to the client indicating ICMP error.

The ICMP error packet should have following format. (https://www.computernetworkingnotes.com/networking-tutorials/icmp-error-messages-and-format-explained.html)
packet: ipv4 header + ICMP header + original-incoming-ipv4 header + first-8-bytes-of-data-in-original-packet

However, it is noticed that 'original-incoming-ipv4 header' contains all 0s. So, the packet can not be decoded by the client.

### Cilium Version

v1.11

### Kernel Version

5.17.11-1rodete2-amd64 #1 SMP PREEMPT Debian 5.17.11-1rodete2 (2022-06-09) x86_64 GNU/Linux

### Kubernetes Version

1.23

### Sysdump

_No response_

### Relevant log output

```shell
Captured ICMP packet looks like this (I have added dummy 14 bytes L2 header at the front). Check the following packet using any packet decoder (e.g. https://hpd.gasmi.net/)

10 30 40 50 60 70 10 30 40 50 60 50 08 00 45 00 00 38 46 47 40 00 40 01 4B 86 0A FD 93 32 0A C8 00 01 03 04 F7 81 00 00 05 7A 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 07 D0 0D 05 05 6E A2 C9
```

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by nnbu**
With the existing code, icmp packet reply can not be decoded because
it has the following format:
ipv4 packet = ipv4 + icmp + orig_ipv4 (has all 0s) + original_data(first 8 bytes)
ipv6 packet = ipv6 + icmpv6 + orig_ipv6 (has all 0s) + original_data(first 64 bytes)

This change adds the missing inner ip header so that packet can be
correctly decoded.
ipv4 packet = ipv4 + icmp + orig_ipv4 + 8 bytes of original data
ipv6 packet = ipv6 + icmpv6 + orig_ipv6 + 64 bytes of original data

Fixes: #21236
Signed-off-by: Nishant Burte nishantburte@gmail.com

### CODE REVIEW

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: The RFC is mentioning that the IP headers of the original packet + 64 bit of the original datagram's data.

Could you add those extra data? They are expected by some equipments to resolve as an example NAT session

https://www.rfc-editor.org/rfc/rfc792.html

**Review by sahid — CHANGES_REQUESTED** [review]
Thanks for this add!

**Comment by ldelossa** [comment]
@nnbu there seems to be quite a few commits which are not related to this PR. Is this intentional?

**Comment by nnbu** [comment]
> @nnbu there seems to be quite a few commits which are not related to this PR. Is this intentional?

No sorry. I am not very familiar with the process. I am trying to correct this.

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: Thanks. I updated the change now.

**Comment by nnbu** [comment]
My tests are failing saying login/authentication failures.

E.g. https://github.com/nnbu/cilium/actions/runs/3026506501

Could someone tell what I am missing?

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: I'm missing something, why do we want this to happen only for this purpose?

**Comment by sahid** [comment]
Thanks, Can I suggest you to add the same kind of support for `dsr_reply_icmp6`?

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: No. The change is applicable in any encap mode. Thanks for bringing it up.

**Comment by nnbu** [comment]
> Thanks, Can I suggest you to add the same kind of support for `dsr_reply_icmp6`?
Yes, I have now added it for dsr_reply_icmp6.

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: Is that necessary to recompute du whole csum for ip hdr, can't you just make the diff with service and pod?

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: As we can never be in a situation with protocol `ICMP` I would suggest a different pattern and `goto drop_err` if  protocol is different from `UDP/TCP`

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: What would happen to the csum of the the ICMP packet, I think if you change the dport it will have to change too, no?

**Review by sahid — CHANGES_REQUESTED** [review]
Thank you for your changes I have some questions for the ipv4 part but they probably have to be addressed for ipv6 too ;-)

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: Existing checksum is already fold'ed using csum_fold. So, I could not find a way to do a diff. Let me know if you have a way.

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: done

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: Right. I have now corrected this issue. I will add a new comment showing ipv4/icmpv4, ipv6/icmpv6 packet responses which has correct checksum.

**Comment by nnbu** [comment]
This is the ipv4 response of the form: ipv4 + ipcmv4 + inner_ipv4 + 8 bytes. Check using any packet decoder like https://hpd.gasmi.net/

00 50 56 82 78 82 00 50 56 82 3B FB 08 00 45 00 00 38 81 EB 40 00 40 01 C3 62 0A 01 01 85 15 00 D4 F1 03 04 19 A5 00 00 05 D4 45 00 05 D5 81 EB 00 00 40 11 1F C5 15 00 D4 F1 15 00 D4 76 07 D0 0D 03 05 C1 C2 EE

This is the ipv6 response of the form: ipv6 + ipcmv6 + inner_ipv6 + 64 bytes. Check using any packet decoder like https://hpd.gasmi.net/

00 50 56 82 78 82 00 50 56 82 3B FB 86 DD 60 0E 7C 41 00 70 3A 40 26 20 00 00 10 00 26 31 00 03 00 10 00 03 01 22 26 20 00 00 10 00 26 31 02 50 56 FF FE 82 78 82 02 00 9B FF 00 00 05 C4 60 0E 7C 41 05 9D 11 40 26 20 00 00 10 00 26 31 02 50 56 FF FE 82 78 82 26 20 00 00 10 00 26 31 00 03 00 10 00 02 00 02 07 D0 0D 03 05 9D 00 00 35 4D BF 95 43 C4 BE 62 EB 5D 2E 23 DF E4 E3 69 00 94 19 6B E5 40 CF B6 31 CA 79 64 F7 77 8D 76 42 1A 40 A1 2F 60 BE EC B9 C1 90 A1 90 66 A4 70 3F 2F C6 7D FA D5 77 03

File: bpf/lib/nodeport.h
Comment (sahid) [reply bpf/lib/nodeport.h]: I see, in such case you will have to unfold and add the diff. I'm just not sure what is the preference in such case.

Could you add a comment saying something like:

```
/* We do recompute the whole checksum here.
 * Another way would be to unfold checksum and
 * then do the math adding the diff. */
```

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h:268]: You've changed the way that we determine whether the packet is too big. I think that makes sense but I think we should have a bug report for this as it looks like a bug no?

Even if we don't want to report a bug. You changed the way we compute size for ipv6 but not for ipv4, any reason that I may have missed?

Comment (nnbu) [reply bpf/lib/nodeport.h:268]: I added the comment about this change elsewhere. 
https://github.com/nnbu/cilium/commit/ea404982e00fbd835c07f3b5f235a0e869b8094a#r84022786

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: Done.

**Comment by nnbu** [comment]
@sahid Could you approve the request if all the queries are answered satisfactorily?

**Comment by sahid** [comment]
> @sahid Could you approve the request if all the queries are answered satisfactorily?

Sure, sorry @nnbu I was off the last couple of days. I will have look shortly :-)

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: nit: Could you find a better naming than `tmp` for this buffer storing original ip hdr + proto?

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h]: Here you compute the checksum based on the original dgram, that one has the dest port pointing to the pod, right? Should not we have to set it to port port prior to start computing checksum? 

And I think we have the same kind of issue for ipv4

**Review by sahid — CHANGES_REQUESTED** [review]

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: Not really. I update the dport first on line number 482 (for ipv6) and 1584 (for ipv4) and then calculate the checksum. I verified the checksum correctness as this comment

https://github.com/cilium/cilium/pull/21234#issuecomment-1250005935

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h]: done

File: bpf/lib/nodeport.h
Comment (sahid) [reply bpf/lib/nodeport.h]: Yes, my mistake, thank you.

File: bpf/lib/nodeport.h
Comment (sahid) [inline bpf/lib/nodeport.h:1405]: As `tmp` has the full and well updated headers I'm wondering whether it would have been better to call `ctx_store_byte()` only one time with it?

**Review by sahid — APPROVED** [review]
Sounds reasonable from my POV, I'm not sure about how tests should be considered at that point for this part of code?

Even If I have one question regarding how we store the packet. The logic is right, I'm approving the change, hopping that you will get attraction from other contributors.

**Comment by sahid** [comment]
Hello @nnbu , checkpatch is not passed, you will have to fix some cosmetic points.

**Comment by nnbu** [comment]
> Hello @nnbu , checkpatch is not passed, you will have to fix some cosmetic points.

Thank you. `make -C bpf checkpatch` now passes.

File: bpf/lib/nodeport.h
Comment (nnbu) [reply bpf/lib/nodeport.h:1405]: It would work. I code it this way to improve readability and making intentions clear.

**Comment by nnbu** [comment]
@sayboras The test failures look to be unrelated to my change. Could you suggest the next steps please?

**Comment by sayboras** [comment]
:wave: we will need one review from datapath team 

cc @cilium/sig-datapath

File: bpf/lib/nodeport.h
Comment (YutaroHayakawa) [inline bpf/lib/nodeport.h]: We have an indentation miss here.

File: bpf/lib/nodeport.h
Comment (YutaroHayakawa) [inline bpf/lib/nodeport.h]: I think we shouldn't rely on the metadata here. It can be accidentally broken by someone outside of this function. Instead, I think it's better to pass `&addr` here.

https://github.com/cilium/cilium/blob/04ecee7fb453a8b480ed9ba2b0f0a2389aeeda26/bpf/lib/nodeport.h#L573

File: bpf/lib/nodeport.h
Comment (YutaroHayakawa) [inline bpf/lib/nodeport.h]: Same as above. I think we should pass `port` here.

https://github.com/cilium/cilium/blob/04ecee7fb453a8b480ed9ba2b0f0a2389aeeda26/bpf/lib/nodeport.h#L573

File: bpf/lib/nodeport.h
Comment (YutaroHayakawa) [inline bpf/lib/nodeport.h]: Same as IPv6 comment

File: bpf/lib/nodeport.h
Comment (YutaroHayakawa) [inline bpf/lib/nodeport.h]: Same as IPv6 comment

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Overall logic looks good. There are some nits. At least the indentation miss should be fixed. Rest are non-blocker.

File: bpf/lib/nodeport.h
Comment (NikAleksandrov) [inline bpf/lib/nodeport.h]: nit: please keep the multiline comment format consistent with the rest of the file, e.g.:
```
/* first line
 * second line
 */
```

File: bpf/lib/nodeport.h
Comment (NikAleksandrov) [inline bpf/lib/nodeport.h]: same nit about the multiline comment format consistency (see previous for details)

**Review by YutaroHayakawa — APPROVED** [review]
Now, LGTM.

**Comment by nnbu** [comment]
@YutaroHayakawa There are 2 tests failing, but they don't look related to my changes. e.g. one of the failures is `663 Error: unknown flag: --backend-weights`

Anything to be from my side?

Also, waiting for results on 4 tests `k8s-.*-kernel-.*`. Not sure if they take more time or did not run at all.

**Comment by nnbu** [comment]
@YutaroHayakawa I see 2 test failures. Looking for some pointers.

1. Travis CI - This build failure happens only on amd64. The error looks to be unrelated to the changes
```
gpg: can't open 'clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz.sig': No such file or directory
gpg: verify signatures failed: No such file or directory
ERROR: Failed to verify clang+llvm-10.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz
```

2. The other cluster mesh failure should not be encountered due to my code. Is the test flaky?
```
❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --connect-timeout 5 --output /dev/null http://echo-other-node:8080" failed: command terminated with exit code 28
  ℹ️  curl output:
  curl: (28) Connection timeout after 5001 ms
:0 -> :0 = 000
```

**Comment by YutaroHayakawa** [comment]
Maybe you retried 1. and succeeded?

For 2. it seems it is a known flake (https://github.com/cilium/cilium/issues/21655). And I agree that it is not related to your changes. I think we can ignore that.

**Comment by nnbu** [comment]
Travis CI build now passed after it was restarted.

**Comment by nnbu** [comment]
@YutaroHayakawa
Can we also cherry-pick this bug fix in v1.11?

**Comment by YutaroHayakawa** [comment]
Here we have a backport criteria of the current stable - 1 version. I think this change meets the criteria `Major bug fixes relevant to the correct operation of Cilium`. Let me put backport tags for v1.11 and v1.10.

https://docs.cilium.io/en/v1.12/contributing/release/backports/#backport-criteria-for-x-y-1-z-and-x-y-2-z

**Comment by pchaigno** [comment]
> I think this change meets the criteria `Major bug fixes relevant to the correct operation of Cilium`. Let me put backport tags for v1.11 and v1.10.

If we end up having any conflict while backporting, I'd be in favor of dropping it from backports. Calling this a major bug is a bit much IMO, especially without a description of the exact impact this bug can have.

**Comment by YutaroHayakawa** [comment]
@pchaigno Then, how should @nnbu act when backport is dropped? Make backport PR by themselves?

**Comment by sahid** [comment]
I tend to agree with Paul, looks like this sounds more like a new support than a major bug fix. We should be careful on backporting  changes in a stable branch, even if I would be glad to see this support in early releases.

**Comment by qmonnet** [comment]
Backport looked non-trivial, so I've kept it out of the regular backport round for 1.10, 1.11.

**Comment by qmonnet** [comment]
I also [removed it](https://github.com/cilium/cilium/pull/21809#discussion_r1005730148) from the current round of 1.12 backports.

**Comment by pchaigno** [comment]
We've removed the backport of this fix from v1.12 because it was causing verifier errors (`invalid size of register spill`) there. If you need it backported, please open a backport PR.

**Pull request merged**
